### PR TITLE
Don't require `allow_internal_unstable` unless `staged_api` is enabled.

### PR DESCRIPTION
**Pull request by ecstatic-morse**
#63770 changed `qualify_min_const_fn` to require `allow_internal_unstable` for *all* crates that used an unstable feature, regardless of whether `staged_api` was enabled or the `fn` that used that feature was stably const. In practice, this meant that every crate in the ecosystem that wanted to use nightly features added `#![feature(const_fn)]`, which skips `qualify_min_const_fn` entirely.

After this PR, crates that do not have `#![feature(staged_api)]` will only need to enable the feature they are interested in. For example, `#![feature(const_if_match)]` will be enough to enable `if` and `match` in constants. Crates with `staged_api` (e.g., `libstd`) require `#[allow_internal_unstable]` to be added to a function if it uses nightly features unless that function is also marked `#[rustc_const_unstable]`. This prevents proliferation of `#[allow_internal_unstable]` into functions that are not callable in a `const` context on stable.

r? @oli-obk (author of #63770)
cc @Centril

### CODE REVIEW

File: src/librustc_mir/transform/qualify_min_const_fn.rs [hunk 1]
```diff
@@ -281,8 +281,21 @@ fn check_place(
     Ok(())
 }
 
-/// Returns whether `allow_internal_unstable(..., <feature_gate>, ...)` is present.
+/// Returns `true` if the given feature gate is allowed within the function with the given `DefId`.
 fn feature_allowed(tcx: TyCtxt<'tcx>, def_id: DefId, feature_gate: Symbol) -> bool {
+    // All features require that the corresponding gate be enabled.
```
Comment (Centril) [inline src/librustc_mir/transform/qualify_min_const_fn.rs]:
```suggestion
    // All features require that the corresponding gate be enabled,
    // even if the function has `#[allow_internal_unstable(the_gate)]`.
```

**Comment by Centril** [comment]
r=me with comment ^-- added.

**Comment by Centril** [comment]
If you can follow up with changes to the rustc guide that'd be great.

**Comment by ecstatic-morse** [comment]
I'll r+ on your behalf once this passes CI. There might be some bootstrap problems now that the `feature` is required to be at the crate root as well as in `allow_internal_unstable`.

**Comment by nbdd0121** [comment]
Note that currently there is a workaround using macros:
```rust
#![feature(allow_internal_unstable)]

#[allow_internal_unstable(const_transmute)]
macro_rules! transmute {
    ($e: expr) => {
        std::mem::transmute($e)
    };
}

#[allow_internal_unstable(wrapping_next_power_of_two)]
macro_rules! wrapping_next_power_of_two {
    ($e: expr) => {
        $e.wrapping_next_power_of_two()
    };
}


fn g(a: u64) -> u64 {
    wrapping_next_power_of_two!(a)
}

const fn h(a: u64) -> i64 {
    unsafe { transmute!(a) }
}
```

**Pull request merged**